tests/functional/test_configured_endpoint_url.py

```diff
@@ -0,0 +1,283 @@
+
+
+def dict_to_ini_section(ini_dict, section_header):
+    profile_str = f'[{section_header}]\n'
+    for key, value in ini_dict.items():
+        if isinstance(value, dict):
+            profile_str += f"{key} =\n"
+            for new_key, new_value in value.items():
+                profile_str += f"  {new_key}={new_value}\n"
+        else:
+            profile_str += f"{key}={value}\n"
+    return profile_str + "\n"
+
+
+def create_cases(test_suite):
+    profiles = test_suite['profiles']
+    service_sections = test_suite['services']
+    environments = test_suite['environments']
+    client_configs = test_suite['client_configs']
+    tests = test_suite['endpointUrlTests']
+
+    for test in tests:
+        new_test = test.copy()
+
+        augmented_data = {}
+        environment_name = test.get('environment', None)
+        environment = environments.get(environment_name, {})
+        augmented_data['environment'] = environment
+
+        client_config_name = test.get('client_config', None)
+        client_config = client_configs.get(client_config_name, {})
+        augmented_data['client_config'] = client_config
+
+        profile_name = test.get('profile', None)
+        profile = profiles.get(profile_name, {})
+        augmented_data['profile'] = profile
+
+        service_section_name = profile.get('services', None)
+        service_section = service_sections.get(service_section_name, {})
+        augmented_data['services'] = service_section
+
+        new_test['augmented_data'] = augmented_data
+        yield new_test
+
+
+def _normalize_endpoint(url):
+    split_endpoint = urlsplit(url)
+    actual_endpoint = f"{split_endpoint.scheme}://{split_endpoint.netloc}"
+    return actual_endpoint
+
+
+def assert_endpoint_url_used(client, expected_endpoint_url):
+    assert client.meta.endpoint_url == expected_endpoint_url
+
+
+def assert_endpoint_url_used_for_operation(
+    client, expected_endpoint_url, operation=None, params=None
+):
+    http_stubber = ClientHTTPStubber(client)
+    http_stubber.start()
+    http_stubber.add_response()
+
+    # Call an operation on the client
+    getattr(client, operation)(**params)
+
+    assert (
+        _normalize_endpoint(http_stubber.requests[0].url)
+        == expected_endpoint_url
+    )
+
+
+def idfn(test_case):
+    return test_case['name']
+
+
+def setup_configuration_environment(test_case):
+    profile_str = dict_to_ini_section(
+        test_case['augmented_data']['profile'],
+        section_header=f"profile {test_case['profile']}",
+    )
+    services_section_name = test_case['augmented_data']['profile'].get(
+        'services', None
+    )
+    service_section_str = dict_to_ini_section(
+        test_case['augmented_data']['services'],
+        section_header=f'services {services_section_name}',
+    )
+    test_case['augmented_data']['profile_string'] = (
+        profile_str + service_section_str
+    )
+
+
+def get_env_config_file_patched_client(session, test_case):
```

Comment:
Let's make the following updates to this function:
1. Rename it to `create_client()`. The added descriptions to the function name make it hard to follow.
2. Turn the `create_client()` function into a pytest fixture that yields back a factory to create clients. This will allow us to keep the temporary file and environment patch around for the length of the test instead of have it end when the client is returned
3. Create the session as part of the factory. This will allow us to avoid patching the `configprovider` module

So in the end, I'd imagine it would look roughly like this:
```python
@pytest.fixture
def create_client(tmp_path):
    tmp_config_file = tmp_path / config
    environ = {'AWS_CONFIG_FILE': tmp_config_file}
    def _do_create_client(service, profile, client_args=None, config_file_contents=None, environment=None):
         environ.update(environment)
         # Write contents to tmp config file here
         session = botocore.session.Session(profile=profile)
         return client = session.create_client(service, **client_args)

    with mock.patch('os.envrion', environ)
        yield _do_create_client
```